Update libmesh, mfem, nlohmann json; remove vtk patch; minor doc fixes

roystgnr opened the pull request:
Closes #30372

- More rigorous implementation and verification of `MeshBase::clone()`.  This can fix issues that arise when certain `MOOSE` mesh generators accept input meshes which are incompletely prepared and have been cloned for use in multiple mesh generators.
- Improved support for vector-valued elements on meshes with 2D elements outside the XY plane in 3D space
- Initial support for `Polyhedron` type elements, with one basic subclass and hand-coded unit tests.  Critical support for mesh generation and I/O is still forthcoming.
- The `Elem::set_node()` setter which returns a reference has been deprecated and replaced by a version which takes the value to set.  This provides compatibility with new Polyhedron support.
- Added `Elem::type_to_dim_map` and `Elem::type_to_default_order_map`, enabling code to query the dimension and/or default order of an `ElemType` without actually building an element.
- `MeshBase::write()` is now a `const` member
- `Elem::disconnected_clone()` can now clone elements which have not yet been given a valid `unique_id()`.
- Autodetection of slate use in PETSc, to initialize MPI with the `MPI_THREAD_MULTIPLE` setting for slate compatibility. 
- Backported NetCDF fix for compatibility with C23 C compilers
- `solid_angle()` function for evaluating the solid angle defined by three vectors
- Code refactoring for clarity and simplicity:

  - Orientation-determining code for advanced finite element types
  - `Tet::choose_diagonal()` for tetrahedron refinement
  - More local variables declared `const` as appropriate
  - Unit tests are easier to make compatible with `--enable-complex` builds
    
- Fixes for compiler `-fsanitize-integer` warnings
- Many fixes for libMesh builds with `Real` defined to be other than `double`.  Many more libMesh capabilities such as higher-order quadrature rules are no longer truncated to 64-bit or 80-bit accuracy when used with 128-bit-`Real` builds.
- Avoid poor triangulator behavior when given an inconsistent boundary input mesh 
- Performance logging of `n_connected_components()` mesh tool

roystgnr committed:
Newsletter entries for libMesh update

roystgnr committed:
Update libMesh submodule

- Added a `System::parameters` object, which can be used to
  selectively override `EquationSystems::parameters`
- Improved support for vector-valued elements on meshes with 2D
  elements outside the XY plane in 3D space
- Initial support for `Polyhedron` type elements, with one basic
  subclass and hand-coded unit tests.  Critical support for mesh
  generation and I/O is still forthcoming.
- The `Elem::set_node()` setter which returns a reference has been
  deprecated and replaced by a version which takes the value to set.
  This provides compatibility with new Polyhedron support.
- Added `Elem::type_to_dim_map` and `Elem::type_to_default_order_map`,
  enabling code to query the dimension and/or default order of an
  `ElemType` without actually building an element.
- Added `NumericVector::print_matlab()` and
  `NumericVector::read_matlab()` I/O methods
- `MeshBase::write()` is now a `const` member
- `Elem::disconnected_clone()` can now clone elements which have not
  yet been given a valid `unique_id()`.
- Autodetection of slate use in PETSc, to initialize MPI with the
  `MPI_THREAD_MULTIPLE` setting for slate compatibility.
- `solid_angle()` function for evaluating the solid angle defined by
  three vectors
- Code refactoring for clarity and simplicity:

  - Orientation-determining code for advanced finite element types
  - `Tet::choose_diagonal()` for tetrahedron refinement
  - More local variables declared `const` as appropriate
  - Unit tests are easier to make compatible with `--enable-complex`
    builds

- Removed workaround for overzealous gcc 8.1 warnings about gzstream
  headers; newer compilers including gcc 8.5 do not emit warnings.
- Avoid poor triangulator behavior when given an inconsistent boundary
  input mesh
- Performance logging of `n_connected_components()` mesh tool
- Many bug fixes:

  - More rigorous implementation and verification of
    `MeshBase::clone()`.  This fixes issues that can arise when
    certain `MOOSE` mesh generators accept input meshes which are
    incompletely prepared and have been cloned for use in multiple mesh
    generators.
  - `UnstructuredMesh::copy_nodes_and_elements()` no longer copies
    other metadata such as preparation policy settings from its source
    mesh.  That copying has been moved to copy constructors which need
    it, and no longer occurs during mesh stitching.  This fixes
    partitioning failures in a reactor module mesh generator.
  - `UnstructuredMesh::copy_nodes_and_elements()` is more careful
    about marking its target as unprepared if any combination of
    inputs might cause it to be unprepared
  - Fix for automatic interior-parents detection on refined meshes
  - Backported NetCDF fix for compatibility with C23 C compilers
  - Fixes for compiler `-fsanitize-integer` warnings
  - Many fixes for libMesh builds with `Real` defined to be other than
    `double`.  Many more libMesh capabilities such as higher-order
    quadrature rules are no longer truncated to 64-bit or 80-bit
    accuracy when used with 128-bit-`Real` builds.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 8d1b4c43d1ab2283d4702aa86cf040d2ae2ec5c0
+Subproject commit d932b6b91da2fac62509af14c56e169f3317f12b
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2865108/), step Docs: sync website on 57519e5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30485/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2865155/), step Generate coverage on 57519e5 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30485/coverage/xfem)

</details>



This comment will be updated on new commits.

nmnobre committed:
Remove vtk patch

refs #30372

roystgnr committed:
Use new set_node(n,node) overload in framework

roystgnr committed:
Use new set_node(n,node) overload in modules

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2868291/), step Versioner verify on 1db5136 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | e0eb0da | 6cce7c7   | 2025.04.22 build 0 | 2025.05.09 build 0 |
| moose-dev | CHANGE   | 8a3a8dd | 4bc20a4   | 2025.04.22         | 2025.05.09         |

lindsayad commented:
Once this goes in the [WIP Cuda container](https://civet.inl.gov/job/2876212/) should pass on https://github.com/idaholab/moose/pull/28114 since I need to be able to use `MPI_THREAD_MULTIPLE`

nmnobre committed:
Fix minor typo in GeneratedMeshGenerator's param description

nmnobre committed:
Update MFEM

nmnobre committed:
Fix docs for MFEM curl curl kernel and integrated bc

nmnobre committed:
Ignore VTK version when verifying MFEM test results

nmnobre committed:
Make sure vanilla nlohmann is included

refs #30372

nmnobre committed:
Update nlohmann to version 3.12.0

refs #30372

nmnobre committed:
Check for conduit files in outputs test

loganharbour committed:
Add missing dependencies for mfem in moose-dev

loganharbour commented:
@roystgnr I just realized I missed some dependencies when adding mfem a ways back. When you get to updating this, can you please incorporate https://github.com/idaholab/moose/commit/28c52702a644b3ed5d8c3584e2f6a2b4072af814 from my fork immediately before the version bumps?

roystgnr committed:
More libMesh update news

Plus some refactoring for categorization now that the "bug fixes" list
has grown.

loganharbour committed:
Bump libmesh-vtk, libmesh, and moose-dev

loganharbour committed:
Update versioner_hashes

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 8d1b4c43d1ab2283d4702aa86cf040d2ae2ec5c0
+Subproject commit 0846b207f9fd5b70f5f5088684bbfe3d1a7785da
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2896091/), step Versioner verify on 88eb0c0 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner influential files

Found 57 influential files, 7 changed, 2 added, 1 removed

| package     | status   | file                                              |
|-------------|----------|---------------------------------------------------|
| libmesh-vtk | CHANGE   | conda/libmesh-vtk/meta.yaml                       |
| libmesh-vtk | REMOVED  | conda/libmesh-vtk/vtk9.4.2_mr11854_nlohmann.patch |
| libmesh     | CHANGE   | apptainer/libmesh.def                             |
| libmesh     | CHANGE   | conda/libmesh/conda_build_config.yaml             |
| libmesh     | CHANGE   | conda/libmesh/meta.yaml                           |
| libmesh     | CHANGE   | libmesh                                           |
| moose-dev   | CHANGE   | conda/moose-dev/conda_build_config.yaml           |
| moose-dev   | CHANGE   | conda/moose-dev/meta.yaml                         |
| moose-dev   | NEW      | framework/contrib/conduit                         |
| moose-dev   | NEW      | framework/contrib/mfem                            |

### Versioner versions

Found 10 packages, 3 changed, 0 failed

| package     | status   | hash    | to hash   | version            | to version         |
|-------------|----------|---------|-----------|--------------------|--------------------|
| libmesh-vtk | CHANGE   | 59e681e | 042bd0b   | 9.4.2 build 1      | 9.4.2 build 2      |
| libmesh     | CHANGE   | e0eb0da | 982ea19   | 2025.04.22 build 0 | 2025.05.23 build 0 |
| moose-dev   | CHANGE   | 24fdce3 | 92d3c2e   | 2025.05.20         | 2025.05.23         |

roystgnr reviewed (commented):
Still happy with the libMesh part of this, at least.

loganharbour reviewed (approved):
Approving of changes that are not my own.

nmnobre commented:
> Still happy with the libMesh part of this, at least.

Do you think it's worth mentioning the caching changes/fixes on the changelog/newsletter? I'm aware that I'm probably the only one who's been affected by it, but in case I messed up, at least it's logged (beyond commit messages of course).

nmnobre commented:
> Approving of changes that are not my own.

Thanks for putting it all together @loganharbour. :)
And thanks for noting my changes closed issue #30372, missed that.

nmnobre reviewed (approved):
> @cticenhour if you could give a second set of eyes on the package bumps please

I'm no Casey, but LGTM (compared with what I had done, and I don't think you missed anything). I also trust @roystgnr used regex replacement for the `set_node()` syntax changes?

roystgnr commented:
> Do you think it's worth mentioning the caching changes/fixes on the changelog/newsletter? I'm aware that I'm probably the only one who's been affected by it, but in case I messed up (of course I didn't 😇), at least it's logged (beyond commit messages of course).

I definitely do, but those didn't make it into this submodule update, sorry.  There was a regression in the --disable-amr case, I didn't get the fix in until this morning, and so your changes hadn't reached master by the time I updated the PR.

> I also trust @roystgnr used regex replacement for the set_node() syntax changes?

Regex for all the changes to set_node() calls that fit on one line, but I didn't trust myself to figure out multiline regex, so for the others I manually did a search/join/regex-application on each, and it's not impossible that I typoed something.  That's like 3 characters in vim, but it's 3 characters that I had to hit a hundred times while also manually watching to see if anything unexpected was happening.

milljm reviewed (changes requested):
Conda stuff looks good to me...
